libindy/src/commands/metrics.rs

```diff
@@ -12,18 +29,585 @@ const OPENED_WALLET_IDS_COUNT: &str = "opened_wallet_ids_count";
+    PaymentsCommandBuildPaymentReq,
+    PaymentsCommandBuildPaymentReqAck,
+    PaymentsCommandParsePaymentResponse,
+    PaymentsCommandParsePaymentResponseAck,
+    PaymentsCommandAppendTxnAuthorAgreementAcceptanceToExtra,
+    PaymentsCommandBuildMintReq,
+    PaymentsCommandBuildMintReqAck,
+    PaymentsCommandBuildSetTxnFeesReq,
+    PaymentsCommandBuildSetTxnFeesReqAck,
+    PaymentsCommandBuildGetTxnFeesReq,
+    PaymentsCommandBuildGetTxnFeesReqAck,
+    PaymentsCommandParseGetTxnFeesResponse,
+    PaymentsCommandParseGetTxnFeesResponseAck,
+    PaymentsCommandBuildVerifyPaymentReq,
+    PaymentsCommandBuildVerifyPaymentReqAck,
+    PaymentsCommandParseVerifyPaymentResponse,
+    PaymentsCommandParseVerifyPaymentResponseAck,
+    PaymentsCommandGetRequestInfo,
+    PaymentsCommandSignWithAddressReq,
+    PaymentsCommandSignWithAddressAck,
+    PaymentsCommandVerifyWithAddressReq,
+    PaymentsCommandVerifyWithAddressAck,
+    // CacheCommand
+    CacheCommandGetSchema,
+    CacheCommandGetSchemaContinue,
+    CacheCommandGetCredDef,
+    CacheCommandGetCredDefContinue,
+    CacheCommandPurgeSchemaCache,
+    CacheCommandPurgeCredDefCache,
+    // MetricsCommand
+    MetricsCommandCollectMetrics,
+    // Exit
+    Exit,
+}
+
+
 pub enum MetricsCommand {
     CollectMetrics(Box<dyn Fn(IndyResult<String>) + Send>)
 }
 
+const COMMANDS_COUNT: usize = MetricsStorage::commands_count();
+
+pub struct MetricsStorage {
+    queue_count: [u64; COMMANDS_COUNT],
+    queue_duration: [u64; COMMANDS_COUNT],
+
+    execute_count: [u64; COMMANDS_COUNT],
+    execute_duration: [u64; COMMANDS_COUNT],
+}
+
+impl MetricsStorage {
+    pub fn new() -> MetricsStorage {
+        MetricsStorage {
+            queue_count: [u64::MIN; COMMANDS_COUNT],
+            queue_duration: [u64::MIN; COMMANDS_COUNT],
+
+            execute_count: [u64::MIN; COMMANDS_COUNT],
+            execute_duration: [u64::MIN; COMMANDS_COUNT],
+        }
+    }
+
+    pub fn cmd_left_queue(&mut self, command_index: usize, duration: u64) {
+        // TODO: change command_index: usize to command: &Command
+        self.queue_count[command_index] += 1;
+        self.queue_duration[command_index] += duration;
+    }
+
+    pub fn cmd_executed(&mut self, command_index: usize, duration: u64) {
+        // TODO: change command_index: usize to command: &Command
+        self.execute_count[command_index] += 1;
+        self.execute_duration[command_index] += duration;
+    }
+
+    pub fn cmd_index(cmd: &Command) -> usize {
+        let idx: CommandIndexes = cmd.into();
+        idx as usize
+    }
+    pub fn cmd_name(index: usize) -> String {
+        CommandIndexes::from(index).to_string()
+    }
+    const fn commands_count() -> usize {
+        CommandIndexes::VARIANT_COUNT
+    }
+}
+
 pub struct MetricsCommandExecutor {
-    wallet_service: Rc<WalletService>
+    wallet_service: Rc<WalletService>,
+    pub metric_storage: MetricsStorage,
```

Comment:
I'm not sure about practice of using public structure members...

Reply:
I think you're right. Would it be better to make the storage getter or wrappers over the storage functions the safest method?